src/osd/PG.cc

```diff
@@ -6034,6 +6037,85 @@ void PG::update_store_on_load()
   }
 }
 
+
+void PG::_delete_some()
+{
+  dout(10) << __func__ << dendl;
+
+  // this ensures we get a valid result.  it *also* serves to throttle
+  // us a bit (on filestore) because we won't delete more until the
+  // previous deletions are applied.
+  osr->flush();
```

Comment:
hmm, that's a good point.  i'm not sure what a better way to throttle would be, though.  options are:

- wait for commit to queue next delete.  this will be pretty slow if the chunk size is small; large chunk sizes mean more interference with client io.
- wait for onreadable -- this won't work with bluestore, which is immediately readable (or at least claims to be).
- ???

Maybe we want to do something clever here, like make each op accumulate some *read* work (to keep it small and fast), and after many such DeleteSome passes queue up the write transaction and wait for commit?

Reply:
added todo

Reply:
We can spin up callbacks when the deletes commit, right? Maybe just make a max outstanding number and block (via a DeletingWait state) doing a new chunk until they're done?

Reply:
Hrm, we are triggering the next delete step once onreadable.  We can't change that oncommit or else it will break filestore (unless we keep the flush).  It also means that the flush() is a no-op for filestore on subsequent passes because we waited for onreadable.

For bluestore, onreadable is triggered at queue time, which means that without the flush we go really fast without much (any?) throttling.  Keeping the flush throttles it.

I think the fix is to condition the next step on *both* onreadable and oncommit.  That's what the latest patch does.